client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java

```diff
@@ -415,9 +415,15 @@ ClientConnection getConnection() throws SegmentSealedException {
     void setupConnection() throws ConnectionFailedException {
         if (state.getConnection() == null) {
             CompletableFuture<ClientConnection> newConnection = controller.getEndpointForSegment(segmentName)
-                .thenCompose((PravegaNodeUri uri) -> {
-                    return connectionFactory.establishConnection(uri, responseProcessor);
-                });
+                    .thenCompose((PravegaNodeUri uri) -> {
+                        return connectionFactory.establishConnection(uri, responseProcessor);
+                    }).whenComplete((connection, ex) -> {
+                        if (ex != null) {
```

Comment:
I'm not entirely sure why we're doing this. The signature of this methods expects us to throw and exception when a connection fails, and we are supposed to be throwing the exception via `getAndHandleException`. `getAndHandleException` will throw in the case `newConnection` completes exceptionally.

Is this saying that throwing isn't sufficient and we additionally need to set the state of the connection to failed?